### ISSUE TITLE
Calling wait() on finished threads sometimes crashes

### ISSUE THREAD

**Issue by marcusnaslund** [cogneco/ooc-kean#502]
Right now calling `wait()` and then `free()` on a `Promise` or `Future` which has already completed its task results in memory leaks because of the following if-statement:
```ooc
if (this _state == _PromiseState Unfinished)
	this _thread wait()
```
This is because, with several threads (for example creating more promises in `PromiseTest.ooc`), calling `this _thread wait()` on a finished thread results in segfaults. There is no actual reason for this since the normal procedure would be to call `pthread_join()` in order to synchronize and terminate the other thread. Something appears to be wrong in the `thread` sdk.

**Comment by sebastianbaginski** [comment]
Calling pthread_join on already joined or detached thread is undefined. If i remember correctly, Pthread_join forces the os to context switch to the thread and wait for it to exit its main function. If the thread is done then there is nothing to switch to. I think a segfault is a great way for the os to communicate an error in program logic.
We could have a flag in the thread class to check if its alraedy done and not call pthread_join more than once. Im not sure if we should do that there or in the client classes though.

**Comment by marcusnaslund** [comment]
Yes, that's exactly my recent conclusion as well. :)
I am fixing it in `Promise` so that the `Thread` API is as similar to `pthreads` as possible.

**Comment by marcusnaslund** [comment]
On a related topic @sebastianbaginski - do we really need a `Mutex` in `_ThreadPromise` and `_ThreadFuture`? The only possible conflict is on an individual bit (or two) in `_state`...

**Comment by sebastianbaginski** [comment]
I dont know the implementation details and i dont feel too comfortable with browsing source code on the phone, but if there is any possibility of a race then we should protect with a mutex. That's what they are for

**Comment by marcusnaslund** [comment]
Let's discuss tomorrow when you are here.

**Comment by sebastianbaginski** [comment]
I wanted to write morebut the bloody phone just sent my message in the middle. I will check the implementation tomorrow when i have acces to the computer.it is almost impossible to usethis touchscreen conveniently ;)

**Comment by marcusnaslund** [comment]
Haha, no worries. :) Look here for my latest changes:
https://github.com/marcusnaslund/ooc-kean/blob/promise-collector/source/base/Promise.ooc

**Comment by marcusnaslund** [comment]
Will be resolved by #509 .
